include/geopackage/WKB.hpp

```diff
@@ -0,0 +1,403 @@
+        multipolygon_t
+    >;
+
+    using byte_t = uint8_t;
+    using byte_vector = std::vector<byte_t>;
+
+    /**
+     * projection visitor. applied with boost to project from
+     * cartesian coordinates to WGS84.
+     */
+    struct wgs84;
+
+    // prevent instatiation of this struct
+    wkb() = delete;
+
+    /**
+     * Read WKB from a given buffer
+     * @param[in] buffer byte vector buffer
+     * @return geometry wkb::geometry struct containing the geometry data from the buffer
+     */
+    static geometry read(const byte_vector& buffer);
+
+    static bg::srs::dpar::parameters<> get_prj(uint32_t srid);
+
+  private:
+    /**
+     * Read a WKB point into a cartesian model.
+     * @return point_t 
+     */
+    static point_t read_point(const byte_vector&, int&, uint8_t);
+
+    /**
+     * Read a WKB linestring into a cartesian model.
+     * @return linestring_t 
+     */
+    static linestring_t read_linestring(const byte_vector&, int&, uint8_t);
+
+    /**
+     * Read a WKB polygon into a cartesian model.
+     * @return polygon_t 
+     */
+    static polygon_t read_polygon(const byte_vector&, int&, uint8_t);
+
+    /**
+     * Read a WKB multipoint into a cartesian model.
+     * @return multipoint_t 
+     */
+    static multipoint_t read_multipoint(const byte_vector&, int&, uint8_t);
+
+    /**
+     * Read a WKB multilinestring into a cartesian model.
+     * @return multilinestring_t 
+     */
+    static multilinestring_t read_multilinestring(const byte_vector&, int&, uint8_t);
+
+    /**
+     * Read a WKB multipolygon into a cartesian model.
+     * @return multipolygon_t 
+     */
+    static multipolygon_t read_multipolygon(const byte_vector&, int&, uint8_t);
+};
+
+inline typename wkb::point_t wkb::read_point(const byte_vector& buffer, int& index, uint8_t order)
+{
+    double x, y;
+    utils::copy_from(buffer, index, x, order);
+    utils::copy_from(buffer, index, y, order);
+    return point_t{x, y};
+}
+
+inline typename wkb::linestring_t wkb::read_linestring(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    linestring_t linestring;
+    linestring.resize(count);
+    for (auto& child : linestring) {
+        child = read_point(buffer, index, order);
+    }
+
+    return linestring;
+}
+
+inline typename wkb::polygon_t wkb::read_polygon(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    polygon_t polygon;
+    
+    if (count > 1) {
+        // polygons only have 1 outer ring,
+        // so any extra vectors are considered to be
+        // inner rings.
+        polygon.inners().resize(count);
+    }
+
+    auto outer = read_linestring(buffer, index, order);
+    polygon.outer().reserve(outer.size());
+    for (auto& p : outer) {
+        polygon.outer().push_back(p);
+    }
+
+    for (uint32_t i = 1; i < count; i++) {
```

Comment:
Ah good catch, this might be an off-by-one. `boost::geometry` does some interesting indexing for polygon rings, so there might've been a reason, but I might've also made a mistake here. I'll do some investigating and add a comment/note if it should start at 1. Thanks!

Reply:
You were correct here, I'm not sure why I started at 1 and incremented to count. An incoming commit will resolve this and adds a test for reading polygons with holes, such that the inner rings start at index 0.